### ISSUE TITLE
error: MS-style inline assembly is not available: Unable to find target for this triple (no targets are registered)

### ISSUE THREAD

**Issue by DevSidious** [mono/CppSharp#923]
Fantastic library. However it gets me stuck here:

```
Error parsing 'C:\Folder\header.h'
C:\Program Files (x86)\Windows Kits\10\include\10.0.15063.0\um\winnt.h(949,5): error: MS-style inline assembly is not available: Unable to find target for this triple (no targets are registered)
```

And the culprit:

```
__inline ULONGLONG
NTAPI
Int64ShllMod32 (
    _In_ ULONGLONG Value,
    _In_ DWORD ShiftCount
    )
{
    __asm    {                                         <- line 949
        mov     ecx, ShiftCount
        mov     eax, dword ptr [Value]
        mov     edx, dword ptr [Value+4]
        shld    edx, eax, cl
        shl     eax, cl
    }
}
```
Any ideas on how one could move past this nuisance ?

**Comment by tritao** [comment]
How are you setting up the parser?

It looks like the Clang is not being setup in MS mode.

It should be setup automatically by the driver, but you can try doing it manually by calling `options.SetupMSVC()`.

See https://github.com/mono/CppSharp/blob/27efa2ec871eb79a02c8d3c6ae09287317e8d99a/src/Parser/ParserOptions.cs#L49.

I just noticed we're not actually setting up the Clang target triple in this call, maybe we should, like we do in the CLI driver:

https://github.com/mono/CppSharp/blob/2b5c256514ff8f50b0d90dd8ad1cb1cfb0b32b82/src/CLI/Generator.cs#L42

Try this:

`parserOptions.TargetTriple = "i686-pc-win32-msvc";`

**Comment by tritao** [comment]
Failing that, it might be because of a missing LLVM targets initialization.

@ddobrev I think we might be missing this in our LLVM/Clang initialization code:

```
  // Initialize targets first, so that --version shows registered targets.
  llvm::InitializeAllTargets();
  llvm::InitializeAllTargetMCs();
  llvm::InitializeAllAsmPrinters();
  llvm::InitializeAllAsmParsers();
```

Takem from https://github.com/llvm-mirror/clang/blob/master/tools/driver/cc1_main.cpp#L181

**Comment by DevSidious** [comment]
I'm getting the same result with `parserOptions.TargetTriple = "i686-pc-win32-msvc";` .

[This Stack Overflow thread from 2012](https://stackoverflow.com/questions/13301371/llvm-gcc-asm-to-llvm-in-xcode) mentions `-fms-extensions -fenable-experimental-ms-inline-asm` . However, having never used **Clang** I have no idea if the `__asm {...}` style of inline assembly is supposed to be supported nowadays or not. Couldn't find anything on the subject.

If I temporarily comment out the inline assembly in **winnt.h** I can get rid of the parsing errors, however there must be a more elegant way.

**Comment by ddobrev** [comment]
@DevSidious thank you for your report. Please tell me - have you downloaded C++# or built it from source?

**Comment by tritao** [comment]
@DevSidious They do support it nowadays. The Clang parser calls back into LLVM for parsing the inline assembly, which it fails to do in this case because of no available targets registered. I'm pretty sure the issue is the lack of those initialization lines I mentioned previously.

**Comment by DevSidious** [comment]
@ddobrev 

I used procedure number 2, listed below.

1. With **https://github.com/mono/CppSharp/releases/tag/0.8.13**

```
c:\CppSharp-0.8.13\build>DownloadDeps.bat

c:\CppSharp-0.8.13\build>"premake5" --file=scripts/LLVM.lua download_llvm
Downloading: https://github.com/mono/CppSharp/releases/download/CppSharp/llvm-cad143-windows-vs2015-x86-RelWithDebInfo.7z
"C:\Program Files\7-Zip\7z.exe" x llvm-cad143-windows-vs2015-x86-RelWithDebInfo.7z -ollvm-cad143-windows-vs2015-x86-RelWithDebInfo -y

c:\CppSharp-0.8.13\build>goto quit

c:\CppSharp-0.8.13\build>pause
Press any key to continue . . .
c:\CppSharp-0.8.13\build>GenerateProjects.bat
Build Project Generator:

[0] Clean
[1] Visual C++ 2012
[2] Visual C++ 2013
[3] Visual C++ 2015
[4] Visual C++ 2017
[5] GNU Make

Choice: 4
** Warning: os.is() is deprecated, use 'os.istarget()' or 'os.ishost()'.
   @c:\CppSharp-0.8.13\src\CppParser\premake5.lua(9)

Error: c:/CppSharp-0.8.13/build/LLVM.lua:23: Error finding an LLVM build
Press any key to continue . . .
```

2. With **git clone https://github.com/mono/CppSharp.git**

```
c:\cpp_sharp_temp\CppSharp\build>DownloadDeps.bat

c:\cpp_sharp_temp\CppSharp\build>"premake5" --file=scripts/LLVM.lua download_llvm
Downloading: https://github.com/mono/CppSharp/releases/download/CppSharp/llvm-cad143-windows-vs2017-x86-RelWithDebInfo.7z
"C:\Program Files\7-Zip\7z.exe" x llvm-cad143-windows-vs2017-x86-RelWithDebInfo.7z -ollvm-cad143-windows-vs2017-x86-RelWithDebInfo -y

c:\cpp_sharp_temp\CppSharp\build>goto quit

c:\cpp_sharp_temp\CppSharp\build>pause
Press any key to continue . . .
c:\cpp_sharp_temp\CppSharp\build>GenerateProjects.bat
Build Project Generator:

[0] Clean
[1] Visual C++ 2012
[2] Visual C++ 2013
[3] Visual C++ 2015
[4] Visual C++ 2017
[5] GNU Make

Choice: 4
** Warning: os.is() is deprecated, use 'os.istarget()' or 'os.ishost()'.
   @c:\cpp_sharp_temp\CppSharp\src\CppParser\premake5.lua(9)

Using debug LLVM build: c:/cpp_sharp_temp/CppSharp/build/scripts/llvm-cad143-windows-vs2017-x86-Debug
Using release LLVM build: c:/cpp_sharp_temp/CppSharp/build/scripts/llvm-cad143-windows-vs2017-x86-RelWithDebInfo
** Warning: os.is() is deprecated, use 'os.istarget()' or 'os.ishost()'.
   @..._sharp_temp\CppSharp\src\CppParser\Bindings\premake5.lua(3)

** Warning: the flags value Managed has been deprecated and will be removed.
   Use `clr "On"` instead.
   @...rp_temp\CppSharp\src\CppParser\Bindings\CLI\premake5.lua(11)

Generating CppSharp build configuration file 'BuildConfig.cs'
** Warning: os.is() is deprecated, use 'os.istarget()' or 'os.ishost()'.
   @c:\cpp_sharp_temp\CppSharp\build\Tests.lua(97)

** Warning: the flags value Managed has been deprecated and will be removed.
   Use `clr "On"` instead.
   @c:\cpp_sharp_temp\CppSharp\build\Tests.lua(182)

Building configurations...
Running action 'vs2017'...
Generated vs2017/CppSharp.sln...
Generated ../src/Core/CppSharp.csproj...
Generated ../src/AST/CppSharp.AST.csproj...
Generated vs2017/projects/CppSharp.CppParser.vcxproj...
Generated vs2017/projects/CppSharp.CppParser.vcxproj.user...
Generated vs2017/projects/Std-symbols.vcxproj...
Generated vs2017/projects/Std-symbols.vcxproj.user...
Generated ../src/CppParser/Bindings/CSharp/CppSharp.Parser.CSharp.csproj...
Generated vs2017/projects/CppSharp.Parser.CLI.vcxproj...
Generated vs2017/projects/CppSharp.Parser.CLI.vcxproj.user...
Generated ../src/CppParser/ParserGen/CppSharp.Parser.Gen.csproj...
Generated ../src/Parser/CppSharp.Parser.csproj...
Generated ../src/CLI/CppSharp.CLI.csproj...
Generated ../src/Generator/CppSharp.Generator.csproj...
Generated ../src/Generator.Tests/CppSharp.Generator.Tests.csproj...
Generated ../src/Generator.Tests/CppSharp.Generator.Tests.csproj.user...
Generated ../src/Runtime/CppSharp.Runtime.csproj...
Generated ../tests/CLI/CLI.Gen.csproj...
Generated vs2017/projects/CLI.Native.vcxproj...
Generated vs2017/projects/CLI.Native.vcxproj.user...
Generated vs2017/projects/CLI.CLI.vcxproj...
Generated vs2017/projects/CLI.CLI.vcxproj.user...
Generated ../tests/CLI/CLI.Tests.CLI.csproj...
Generated ../tests/CLI/CLI.Tests.CLI.csproj.user...
Generated ../tests/Common/Common.Gen.csproj...
Generated vs2017/projects/Common.Native.vcxproj...
Generated vs2017/projects/Common.Native.vcxproj.user...
Generated ../tests/Common/Common.CSharp.csproj...
Generated ../tests/Common/Common.Tests.CSharp.csproj...
Generated ../tests/Common/Common.Tests.CSharp.csproj.user...
Generated vs2017/projects/Common.CLI.vcxproj...
Generated vs2017/projects/Common.CLI.vcxproj.user...
Generated ../tests/Common/Common.Tests.CLI.csproj...
Generated ../tests/Common/Common.Tests.CLI.csproj.user...
Generated ../tests/CSharp/CSharp.Gen.csproj...
Generated vs2017/projects/CSharp.Native.vcxproj...
Generated vs2017/projects/CSharp.Native.vcxproj.user...
Generated ../tests/CSharp/CSharp.CSharp.csproj...
Generated ../tests/CSharp/CSharp.Tests.CSharp.csproj...
Generated ../tests/CSharp/CSharp.Tests.CSharp.csproj.user...
Generated ../tests/Encodings/Encodings.Gen.csproj...
Generated vs2017/projects/Encodings.Native.vcxproj...
Generated vs2017/projects/Encodings.Native.vcxproj.user...
Generated ../tests/Encodings/Encodings.CSharp.csproj...
Generated ../tests/Encodings/Encodings.Tests.CSharp.csproj...
Generated ../tests/Encodings/Encodings.Tests.CSharp.csproj.user...
Generated vs2017/projects/NamespacesBase.Native.vcxproj...
Generated vs2017/projects/NamespacesBase.Native.vcxproj.user...
Generated ../tests/NamespacesBase/NamespacesBase.CSharp.csproj...
Generated ../tests/NamespacesDerived/NamespacesDerived.Gen.csproj...
Generated vs2017/projects/NamespacesDerived.Native.vcxproj...
Generated vs2017/projects/NamespacesDerived.Native.vcxproj.user...
Generated ../tests/NamespacesDerived/NamespacesDerived.CSharp.csproj...
Generated ../tests/NamespacesDerived/NamespacesDerived.Tests.CSharp.csproj...
Generated ../tests/NamespacesDerived/NamespacesDerived.Tests.CSharp.csproj.user...
Generated ../tests/StandardLib/StandardLib.Gen.csproj...
Generated vs2017/projects/StandardLib.Native.vcxproj...
Generated vs2017/projects/StandardLib.Native.vcxproj.user...
Generated vs2017/projects/StandardLib.CLI.vcxproj...
Generated vs2017/projects/StandardLib.CLI.vcxproj.user...
Generated ../tests/StandardLib/StandardLib.Tests.CLI.csproj...
Generated ../tests/StandardLib/StandardLib.Tests.CLI.csproj.user...
Generated ../tests/VTables/VTables.Gen.csproj...
Generated vs2017/projects/VTables.Native.vcxproj...
Generated vs2017/projects/VTables.Native.vcxproj.user...
Generated ../tests/VTables/VTables.CSharp.csproj...
Generated ../tests/VTables/VTables.Tests.CSharp.csproj...
Generated ../tests/VTables/VTables.Tests.CSharp.csproj.user...
Generated ../examples/SDL/SDL.csproj...
Done (381ms).
Press any key to continue . . .

```

**Comment by tritao** [comment]
@DevSidious Can you try adding those LLVM lines I mentioned above to https://github.com/mono/CppSharp/blob/master/src/CppParser/Parser.cpp#L71 and see if it solves your issue.

If you need some help to do this, hit me up on our Gitter channel.

**Comment by tritao** [comment]
The error you encountered with compiling via step 1. above, is because `DownloadDeps.bat` downloaded packages for VS2015 instead of VS2017. You should run `DownloadDeps.bat` from a VS command prompt corresponding to the VS version you want to use for compilation. I've clarified this on our documentation.

**Comment by DevSidious** [comment]
@tritao I can confirm that **DownloadDeps.bat** downloads the packages for VS2015 instead of VS2017 although it is run in VS2017 developer command prompt. The DownloadDeps.bat in the release package that is. If I clone the repo, the version in there doesn't misbehave.

Let me see with those LLVM lines.

**Comment by tritao** [comment]
I think @ddobrev found and fixed that `DownloadDeps.bat` bug a few days ago which explains it:

https://github.com/mono/CppSharp/commit/4d177b317dbefe7642f66d004b62722e182d7cf7#diff-d57dbe7e68d12f5a74fb96afd1e027bfL52

**Comment by zillemarco** [comment]
I had the same issue a few days ago and fixed it doing this:

In `CppSharp/src/CppParser/Parser.cpp` I've added `#include <llvm/Support/TargetSelect.h>` and 
```
Parser::Parser(CppParserOptions* Opts) : lib(Opts->ASTContext), opts(Opts), index(0)
{
	llvm::InitializeNativeTarget();              // Added line
	llvm::InitializeNativeTargetAsmParser();     // Added line
}
```

Then I've added all the LLVM libraries to fix the linker error (brute force but I tried to find the real useful ones but it kept failing so I kept them all). In `CppSharp/build/LLVM.lua` I've replaced the LLVM links with (obviously keeping the clang links):

```
      "LLVMAnalysis",
      "LLVMAsmParser",
      "LLVMAsmPrinter",
      "LLVMBinaryFormat",
      "LLVMBitReader",
      "LLVMBitWriter",
      "LLVMCodeGen",
      "LLVMCore",
      "LLVMCoroutines",
      "LLVMCoverage",
      "LLVMDebugInfoCodeView",
      "LLVMDebugInfoDWARF",
      "LLVMDebugInfoMSF",
      "LLVMDebugInfoPDB",
      "LLVMDemangle",
      "LLVMExecutionEngine",
      "LLVMGlobalISel",
      "LLVMInstCombine",
      "LLVMInstrumentation",
      "LLVMInterpreter",
      "LLVMipo",
      "LLVMIRReader",
      "LLVMLibDriver",
      "LLVMLineEditor",
      "LLVMLinker",
      "LLVMLTO",
      "LLVMMC",
      "LLVMMCDisassembler",
      "LLVMMCJIT",
      "LLVMMCParser",
      "LLVMMIRParser",
      "LLVMObject",
      "LLVMObjectYAML",
      "LLVMOption",
      "LLVMOrcJIT",
      "LLVMPasses",
      "LLVMProfileData",
      "LLVMRuntimeDyld",
      "LLVMScalarOpts",
      "LLVMSelectionDAG",
      "LLVMSupport",
      "LLVMSymbolize",
      "LLVMTableGen",
      "LLVMTarget",
      "LLVMTestingSupport",
      "LLVMTransformUtils",
      "LLVMVectorize",
      "LLVMX86AsmParser",
      "LLVMX86AsmPrinter",
      "LLVMX86CodeGen",
      "LLVMX86Desc",
      "LLVMX86Disassembler",
      "LLVMX86Info",
      "LLVMX86Utils",
      "LLVMXRay"
```

I hope this is all because I've made the changes on a non-git CppSharp version I have.

**Comment by ddobrev** [comment]
Hello @zillemarco . I am happy to hear from you and that you still use C++#. We haven't forgotten it was you who contributed our CLI. :)
About the fix, we have a pull request at https://github.com/mono/CppSharp/pull/926 which fixes the problem with less dependencies. However, it crashes our Linux CI for still unclear reasons. Have you tried your code on Linux? Any tests of yours might help solve the problem.
In addition, you said you used a custom copy of C++#. Are these the only changes you've made to it? If not, we'd be happy to take pull requests for any other local fixes you have.

**Comment by zillemarco** [comment]
Hi @ddobrev , Nice to be back I could say :). Sorry about my unfinished CLI but I have my head around a lot of work recently :(.

BTW yes, I've made some other changes indeed regarding some problems compiling one of my projects while targeting MSVC since clang couldn't find some Windows API (I don't remember exactly what it was but I do remember that I saw you've already fixed it).

Unfortunately I didn't have the chance to test my changes on a Linux machine since I don't have one.
